subprojects/core/src/main/java/org/gradle/api/internal/tasks/properties/OutputUnpacker.java

```diff
@@ -107,6 +109,23 @@ private static void resolveOutputFilePropertySpecs(
         boolean locationOnly,
         Consumer<OutputFilePropertySpec> consumer
     ) {
+        if (filePropertyType == OutputFilePropertyType.DIRECTORIES || filePropertyType == OutputFilePropertyType.FILES) {
+            // Eagerly unpack the value, because we need to create one property per root
+            Object unpackedValue = unpackValue(value, locationOnly);
+            if (unpackedValue == null) {
+                return;
+            }
+            resolveCompositeOutputFilePropertySpecs(ownerDisplayName, propertyName, unpackedValue, filePropertyType.getOutputType(), fileCollectionFactory, consumer);
+        } else {
+            FileCollectionInternal outputFiles = fileCollectionFactory.resolving((Callable<Object>) () -> unpackValue(value, locationOnly));
```

Comment:
I don't understand how that will help with finalization: `value` is the thing you want to finalize. Though this change puts it in a file collection as a callable which is not finalizable. How does this help?